tests/risks/test_effect.py

```diff
@@ -398,3 +413,118 @@
+        exposure_col = pd.Series(custom_exposure_values, name=self.exposure_column_name)
+        self.population_view.update(exposure_col)
+
+    def on_time_step_prepare(self, event: Event) -> None:
+        exposure_col = pd.Series(custom_exposure_values, name=self.exposure_column_name)
+        self.population_view.update(exposure_col)
+
+    # noinspection PyAttributeOutsideInit
+    def setup(self, builder: Builder):
+        builder.value.register_value_producer(
+            f"{self.risk.name}.exposure",
+            source=self.get_exposure,
+        )
+
+    def get_exposure(self, index: pd.Index) -> pd.Series:
+        data = pd.Series(custom_exposure_values, index=index)
+        return data
+
+
+def _setup_risk_simulation(
+    config: LayeredConfigTree,
+    plugins: LayeredConfigTree,
+    risk: Union[str, Risk],
+    data: Dict[str, Any],
+) -> InteractiveContext:
+    components = [
+        TestPopulation(),
+        risk,
+        SI("some_disease"),
+        NonLogLinearRiskEffect(risk.name, "cause.some_disease.incidence_rate"),
+    ]
+
+    simulation = InteractiveContext(
+        components=components,
+        configuration=config,
+        plugin_configuration=plugins,
+        setup=False,
+    )
+
+    for key, value in data.items():
+        simulation._data.write(key, value)
+
+    simulation.setup()
+    return simulation
+
+
+def test_non_loglinear_effect(base_config, base_plugins, monkeypatch):
+    risk = CustomExposureRisk("risk_factor.test_risk")
+    monkeypatch.setattr("vivarium_public_health.risks.effect.NUM_RR_EXPOSURE_VALUES", 3)
+    effect = NonLogLinearRiskEffect(
+        "risk_factor.test_risk", "cause.some_disease.incidence_rate"
+    )
+
+    risk_effect_exposures = [1, 2, 5]
+    risk_effect_rrs = [2.0, 2.4, 4.0]
+    rr_data = pd.DataFrame(
+        {
+            "affected_entity": "some_disease",
+            "affected_measure": "incidence_rate",
+            "year_start": 1990,
+            "year_end": 1991,
+            "parameter": risk_effect_exposures,
+            "value": risk_effect_rrs,
+        },
+    )
+    # enforce TMREL of 1
+    tmred = {"distribution": "uniform", "min": 1, "max": 1, "inverted": False}
+
+    data = {
+        f"{risk.name}.relative_risk": rr_data,
+        f"{risk.name}.tmred": tmred,
+        f"{risk.name}.population_attributable_fraction": 0,
+        "cause.some_disease.incidence_rate": 1,
+    }
+
+    base_config.update({"population": {"population_size": 10}})
+    simulation = _setup_risk_simulation(base_config, base_plugins, risk, data)
+
+    pop = simulation.get_population()
+    rate = simulation.get_value("some_disease.incidence_rate")(
+        pop.index, skip_post_processor=True
+    )
+    expected_values = np.interp(
+        custom_exposure_values,
+        risk_effect_exposures,
+        np.array(risk_effect_rrs) / 2,  # RRs get divided by RR at TMREL
+    )
+
+    assert np.isclose(rate.values, expected_values, rtol=0.0000001).all()
```

Comment:
I'd add a test or two that tests some of your errors you anticipated. I think it should be relatively easy to make a dataframe you want to throw a particular error and make a test component and initialize a simulation.

Reply:
Good idea I'll add those.